Sources/_CryptoExtras/AES/Block Function.swift

```diff
@@ -0,0 +1,143 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the SwiftCrypto open source project
+//
+// Copyright (c) 2023 Apple Inc. and the SwiftCrypto project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.md for the list of SwiftCrypto project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import Crypto
+@_implementationOnly import CCryptoBoringSSL
+@_implementationOnly import CCryptoBoringSSLShims
+@_implementationOnly import CryptoBoringWrapper
+import Foundation
+
+extension AES {
+    private static let blockSize = 128 / 8
+
+    /// Apply the AES permutation operation in the encryption direction.
+    ///
+    /// This function applies the core AES block operation to `payload` in the encryption direction. Note that this is
+    /// not performing any kind of block cipher mode, and does not authenticate the payload. This is a dangerous primitive
+    /// that should only be used to compose higher-level primitives, and should not be used directly.
+    ///
+    /// - parameter payload: The payload to encrypt. Must be exactly 16 bytes long.
+    /// - parameter key: The encryption key to use.
+    /// - throws: On invalid parameter sizes.
+    public static func permute<Payload: MutableCollection>(_ payload: inout Payload, key: SymmetricKey) throws where Payload.Element == UInt8 {
```

Comment:
What makes `MutableCollection` the right protocol here?

Reply:
Assuming we want to support in-place modification, `MutableCollection` is the simplest protocol that allows us to do what we want. Specifically, `MutableCollection` adds the `set` accessor to the subscript, so we can modify the elements in the `Collection`. It also has a fast-path available to it for collections that can expose their storage.

Almost all other protocols in this space that you can think of refine the `MutableCollection` protocol in some way: for example, `DataProtocol` refines `MutableCollection`.